Feature/ Embedded entities and lists supports

WojciechMazur opened the pull request:
Changes method for encoding embedded entities, lists and optional values. 

New encoding mechanism is not depended from (prefix) path, it stores each value as Datastore native Values, e.q. LongValue, NullValue (used for Option.empty), ListValue, EntityValue (used to store embedded entities).

1. Lists
Lists of elements previously were saved as <listFieldPath>.<idx>. Such convention did not allowed to create search queries checking if array contains some element. 
New implementation creates stores lists as Datastore native arrays, which allows to use them in queries. 

2. Embedded entities
In previous implementation structure was flatten, it was providing possibility to decode such entities, but it was sub optimal. Also it was not possible to decode embedded entities not saved by mutatus. 

3. Sealed traits 
New implementation saves each sealed trait as embedded entity with two fields: 
- type: full path of encoded sealed trait implementation
- value: encoded inner entity or empty object 
  
Also basic mechanism for batch operations was introduced. Currently is supporting only saving (by upsert) and deletion of entities for collection of entities.

WojciechMazur committed:
Changed serialization of entites to value centric. Fixed invalid serialization of lists and optional values

WojciechMazur committed:
Add suport for batched save/deletion on collection of entities

WojciechMazur committed:
Use datastore Value[_] in simpleDecoder instead of primite type

WojciechMazur committed:
Scalafmt fix

WojciechMazur committed:
Restored method decode(Entity) for backawrd compability, replaced instance of this method for values with decodeValue

WojciechMazur committed:
Fixed sealed traits encoding, introduced _meta field storing typename for sealed traits

propensive commented on src/core/mutatus.scala:
```diff
@@ -294,21 +294,27 @@ object Decoder extends Decoder_1 {
-              subtype.typeclass.decodeValue(entity.getValue("value"))
+    encodedValue => {
+      // Naive approach used when fetching entity not indexed by mutatus
+      def firstSucessfull(value: Value[_]) = {
```
Spelling: `firstSuccessful` or maybe just `firstSuccess`?

propensive commented on src/core/mutatus.scala:
```diff
@@ -294,21 +294,27 @@ object Decoder extends Decoder_1 {
+          Try(entityValue.get.getEntity("_meta").getString("typename"))
+            .map { typeName =>
+              st.subtypes.collectFirst {
+                case subtype if subtype.typeName.full == typeName =>
```
Do we need to use the full typename, or would it be sufficient just to use the class name? I know that it would be possible to have two types in the same file with the same name, inside different objects, both of which inherit from the same sealed trait... but I also think it would be reasonable not to support that case. The benefit would be that the contents of the database would be closer to a "natural" encoding.

One other possibility would be to use the "full" `typeName` only in the case where there would be an ambiguity, which you could determine by checking the set of `subtypes` in the `SealedTrait` type. While that wouldn't be resilient to changes to the data definitions if they were to rename classes... in general these changes will always break things, so it's not worth trying to accommodate them anyway.

WojciechMazur replied:
I think class name on it's own may be better in such case. If package of sealed trait would change it would need as to decode it using naive method of checking all possible case. In case if all implementations of sealed trait are case objects (eq. enums) it would take first successfully decoded... which would be the same for each entity. It may create huge mess in app.

propensive commented on src/core/mutatus.scala:
```diff
@@ -294,21 +294,27 @@ object Decoder extends Decoder_1 {
+
+      encodedValue match {
+        case entityValue: EntityValue =>
+          Try(entityValue.get.getEntity("_meta").getString("typename"))
```
One very minor suggestion: use a `final val` for `"_meta"`, just to avoid the risk of a typo.

propensive reviewed (changes requested):
Basically not much to change, then I can approve and merge it. :)

WojciechMazur committed:
Fixes due to code review
Replaced mutable colletction with Inmutanble
Used constructMondic in decoder
Introduced final vals for field names
Naming and style fixes
